### PR TITLE
Support aeson 2.0

### LINKED ISSUE #27: fails to build with aeson-2.0
```
aeson-combinators              > Building library for aeson-combinators-0.0.5.0..
aeson-combinators              > [1 of 2] Compiling Data.Aeson.Combinators.Decode
aeson-combinators              >               
aeson-combinators              > /tmp/stack-d9db07aa05c3b14b/aeson-combinators-0.0.5.0/lib/Data/Aeson/Combinators/Decode.hs:348:27: error:
aeson-combinators              >     • Couldn't match type: Data.Aeson.KeyMap.KeyMap a
aeson-combinators              >                      with: HS.HashMap Text a
aeson-combinators              >       Expected: Decoder (HS.HashMap Text a)
aeson-combinators              >         Actual: Decoder (Data.Aeson.KeyMap.KeyMap a)
aeson-combinators              >     • In the expression:
aeson-combinators              >         Decoder
aeson-combinators              >           $ \case
aeson-combinators              >               Object xs -> traverse d xs
aeson-combinators              >               val -> typeMismatch "Array" val
aeson-combinators              >       In an equation for ‘hashMapLazy’:
aeson-combinators              >           hashMapLazy (Decoder d)
aeson-combinators              >             = Decoder
aeson-combinators              >                 $ \case
aeson-combinators              >                     Object xs -> traverse d xs
aeson-combinators              >                     val -> typeMismatch "Array" val
aeson-combinators              >     • Relevant bindings include
aeson-combinators              >         d :: Value -> Parser a
aeson-combinators              >           (bound at lib/Data/Aeson/Combinators/Decode.hs:348:22)
aeson-combinators              >         hashMapLazy :: Decoder a -> Decoder (HS.HashMap Text a)
aeson-combinators              >           (bound at lib/Data/Aeson/Combinators/Decode.hs:348:1)
aeson-combinators              >     |         
aeson-combinators              > 348 | hashMapLazy (Decoder d) = Decoder $ \case
aeson-combinators              >     |                           ^^^^^^^^^^^^^^^...
aeson-combinators              >               
aeson-combinators              > /tmp/stack-d9db07aa05c3b14b/aeson-combinators-0.0.5.0/lib/Data/Aeson/Combinators/Decode.hs:357:29: error:
aeson-combinators              >     • Couldn't match type: Data.Aeson.KeyMap.KeyMap a
aeson-combinators              >                      with: HS.HashMap Text a
aeson-combinators              >       Expected: Decoder (HS.HashMap Text a)
aeson-combinators              >         Actual: Decoder (Data.Aeson.KeyMap.KeyMap a)
aeson-combinators              >     • In the expression:
aeson-combinators              >         Decoder
aeson-combinators              >           $ \case
aeson-combinators              >               Object xs -> traverse d xs
aeson-combinators              >               val -> typeMismatch "Array" val
aeson-combinators              >       In an equation for ‘hashMapStrict’:
aeson-combinators              >           hashMapStrict (Decoder d)
aeson-combinators              >             = Decoder
aeson-combinators              >                 $ \case
aeson-combinators              >                     Object xs -> traverse d xs
aeson-combinators              >                     val -> typeMismatch "Array" val
aeson-combinators              >     • Relevant bindings include
aeson-combinators              >         d :: Value -> Parser a
aeson-combinators              >           (bound at lib/Data/Aeson/Combinators/Decode.hs:357:24)
aeson-combinators              >         hashMapStrict :: Decoder a -> Decoder (HS.HashMap Text a)
aeson-combinators              >           (bound at lib/Data/Aeson/Combinators/Decode.hs:357:1)
aeson-combinators              >     |         
aeson-combinators              > 357 | hashMapStrict (Decoder d) = Decoder $ \case
aeson-combinators              >     |                             ^^^^^^^^^^^^^^^...
aeson-combinators              >               
aeson-combinators              > /tmp/stack-d9db07aa05c3b14b/aeson-combinators-0.0.5.0/lib/Data/Aeson/Combinators/Decode.hs:404:26: error:
aeson-combinators              >     • Couldn't match expected type ‘Key’ with actual type ‘Text’
aeson-combinators              >     • In the second argument of ‘(.:)’, namely ‘t’
aeson-combinators              >       In the second argument of ‘(=<<)’, namely ‘v .: t’
aeson-combinators              >       In the expression: d =<< v .: t
aeson-combinators              >     |         
aeson-combinators              > 404 |   Object v -> d =<< v .: t
aeson-combinators              >     |                          ^
aeson-combinators              >               
aeson-combinators              > /tmp/stack-d9db07aa05c3b14b/aeson-combinators-0.0.5.0/lib/Data/Aeson/Combinators/Decode.hs:457:25: error:
aeson-combinators              >     • Couldn't match expected type ‘Text’ with actual type ‘Key’
aeson-combinators              >     • In the first argument of ‘key’, namely ‘txt’
aeson-combinators              >       In the expression: key txt
aeson-combinators              >       In an equation for ‘element’: element (Key txt) = key txt
aeson-combinators              >     |         
aeson-combinators              > 457 | element (Key txt) = key txt
aeson-combinators              >     |                         ^^^
```

### PR DESCRIPTION
**Pull request by fendor**
Adds support for aeson 2.0 and closes #27

The current approach provides full backward compatibility with older aeson versions. So, users with older aeson versions don't have to update anything.

However, users of aeson 2.0 might have to change some code, such as calls to `key` and `at`. It is possible to write backwards and forward compatible code using `fromText` and `toText`, but it is not enforced.

### CODE REVIEW

**Comment by fendor** [comment]
As a very quick and dirty experiment, updating my own projects to aeson 2.0 was trivially possible, mostly due to my use of literals and `OverloadedStrings`.

File: lib/Data/Aeson/Combinators/Compat.hs [hunk 1]
```diff
@@ -0,0 +1,67 @@
+{-# LANGUAGE CPP #-}
+-- |
+-- Module      : Data.Aeson.Combinators.Decode
+-- Copyright   : (c) Marek Fajkus
+-- License     : BSD3
+--
+-- Maintainer  : marek.faj@gmail.com
```
Comment (fendor) [inline lib/Data/Aeson/Combinators/Compat.hs:7]: I volunteered you as the maintainer. Hope that's OK, but I am also fine with my name and email here if you prefer that

File: DocTest.hs [hunk 2]
```diff
@@ -1,6 +1,8 @@
 import Test.DocTest
-main = doctest [ "-isrc"
-               , "-XOverloadedStrings"
+main = doctest [ "-XOverloadedStrings"
+               , "-XCPP"
+               , "-XLambdaCase"
+               , "lib/Data/Aeson/Combinators/Compat.hs"
```
Comment (fendor) [inline DocTest.hs:5]: Doctest doesn't work for me, due to CPP, not sure whether I broke it or it was broken before

**Comment by turboMaCk** [comment]
Sorry I was AFK for a weekend. Great work. You have seemed to approach this the way I would. I will do proper review after work on this.

File: default.nix [hunk 3]
```diff
@@ -21,8 +21,8 @@ let
 in with pkgs; {
   ghc = haskellPackages.aeson-combinators;
   ghcjs = ghcjs.pkgs.haskell.packages.ghcjs.callCabal2nix "aeson-combinators" ./. {};
-  ghc865 = haskell.packages.ghc865.aeson-combinators;
-  ghc883 = haskell.packages.ghc883.aeson-combinators;
   ghc884 = haskell.packages.ghc884.aeson-combinators;
-  ghc8101 = haskell.packages.ghc8101.aeson-combinators;
+  ghc8107 = haskell.packages.ghc8107.aeson-combinators;
+  ghc902 = haskell.packages.ghc902.aeson-combinators;
+  ghc921 = haskell.packages.ghc921.aeson-combinators;
```
Comment (turboMaCk) [inline default.nix]: for some reason even when using these 2 attributes ghc-pkg says I'm using `aeson-1.5.6.0`

Comment (fendor) [reply default.nix]: At least on my nightly:
```
> haskell.packages.ghc921.aeson.version
"1.5.6.0"
```

So, nix is still going to use 1.5.6.0. I can imagine that stackage didn't manage to migrate to aeson 2.0.0.0 yet.

**Comment by turboMaCk** [comment]
It would be great if we could test aeson 2 in CI. However I don't think it's currently present in nixpkgs (unstable). What did you use for testing this? stack nightly or cabal directly?

**Comment by turboMaCk** [comment]
Ok I guess the first step for this is to reproduce build failure in CI. I prepared this PR for it but as you can see it passes because all the sets are using aeson 1 https://github.com/turboMaCk/aeson-combinators/pull/29

**Comment by fendor** [comment]
I used `cabal test --contraint="aeson==1.5.6.0"` and `cabal test` for aeson 2.0.0.0 testing.

**Comment by turboMaCk** [comment]
ok CI confirmed this builds with aeson 2.0. Great work!

**Pull request merged**